Add Telegram support, colors and puzzle solving timestamps

iburakov opened the pull request:
Hi there! Thanks for sharing this code, it was very helpful. 

I've done some further work on it for my own requirements and created a [fork](https://github.com/iburakov/advent-of-code-leaderboard-notifier-telegram) as a separate repo with its own builds for the time being. 

Just suggesting you to take a look, because I have nothing against merging this work back to yours (if you consider it reasonable, of course).

I've simplified writing custom webhook senders and implemented one for Telegram. I've also added colors and event times to notifications, so now they look like this:

<img src='https://github.com/iburakov/advent-of-code-leaderboard-notifier-telegram/assets/5229130/83e0aad9-1b2f-4453-945c-2ae96b83b091' height='300'>

The formatting part is pretty opinionated, yeah. If you don't like it, a way to please everyone would be to accept pattern str in another env var and do a `.format(day=..., part=..., part_emoji=...)`. So, your format becomes `NOTIFICATION_PATTERN="{member} solved day {day} part {part}"`, mine becomes `NOTIFICATION_PATTERN="Day {day}: {member} got {part_emoji} at {when}"`, users can come up with every other thing they like, etc.

So, if you're open to merging it, let me know, because (at least) a README needs some changes for it to be mergeable :) Otherwise feel free to close this PR.

iburakov committed:
Add fork description to README

iburakov committed:
Generalize webhook senders, implement one for Telegram

iburakov committed:
Add exception handling to the loop so sleeping is ensured

iburakov committed:
Strip telegram chat id just in case a whitespace slips

iburakov committed:
Add colors and event times to notifications

iburakov committed:
Describe new features and their usage in a README

iburakov committed:
Fix usage commands in readme

tofran commented:
@iburakov Thank you very much for all the effort! I love it, and thank you for preserving the code-style.

I will have a better look later. But the only thing I will point outright now is the emoji, the colored circle is not very representative - at least in my opinion. I would change it to one of the following (1st, 2nd puzzle):

- ⭐, 🌟
- 🌲, 🎄
- 🌱, 🌲/🎄

And yes, I actually like the `NOTIFICATION_PATTERN` idea. It actually allows for easy i18n that would be useful in my case.

Edit: maybe having two patterns, one for each part will solve everything. It can include the emoji in it.

tofran commented:
Woops the action failed because it is directly pushing the image, I would imagine it would do some sort of linting. Please ignore it. I will fix it in another PR

tofran commented on webhook_senders.py:
```diff
@@ -0,0 +1,70 @@
+import requests
+
+
+class WebhookSender(ABC):
```
I would call this `BaseNotificationSender`.

iburakov replied:
Yeah, that generalizes the concept for even more applications. Ok, let's do it, and I'll try to make sense of everything related like renaming the `WEBHOOK_SENDER` env var to `NOTIFICATION_SENDER` as well.

tofran commented on webhook_senders.py:
```diff
@@ -0,0 +1,70 @@
+        ...
+
+
+class DefaultSender(WebhookSender):
```
.... and this `WebhookSender`

iburakov replied:
yep, makes sense, given the change above

tofran commented on README.md:
```diff
@@ -1,40 +1,68 @@
 Send a webhook notification when someone from an Advent Of Code leaderboard solves a puzzle.
 
-I'm using this to track the progress of a group of friends on Discord.
+This is a fork of [tofran/advent-of-code-leaderboard-notifier](https://github.com/tofran/advent-of-code-leaderboard-notifier) that:
```
As you mentioned this needs work, if you wish, I can help with that. Feel free to enable author contribution permission.

iburakov replied:
yeah, you should have the rights, go ahead :)

~use the [to-be-merged-back](https://github.com/iburakov/advent-of-code-leaderboard-notifier-telegram/tree/to-be-merged-back) branch~ 
nope, use `main` :)

tofran commented on README.md:
```diff
@@ -1,40 +1,68 @@
+`telegram` webhook sender params:
 
+- `WEBHOOK_TELEGRAM_URL`: The URL of your Telegram bot's webhook. Usually it's `https://api.telegram.org/bot<token>/sendMessage`. Ask [BotFather](https://t.me/botfather) for a token.
+- `WEBHOOK_TELEGRAM_CHAT_ID`: comma-separated list of chat IDs to send notifications to. See [Telegram docs](https://core.telegram.org/bots/api#sendmessage) for more info. You can find them using bots like [@JsonDumpBot](https://t.me/JsonDumpBot).
```
Please prefix it with an `S` for plural: `WEBHOOK_TELEGRAM_CHAT_IDS`

iburakov replied:
right, thanks

tofran reviewed (changes requested):
I like it! Thank you for the logging addition.

What do you think of these changes?:

iburakov commented:
That's nice to hear you like it, thanks! Great idea about distinct patterns for both notifications, that'll be much more flexible, indeed.

Here are my plans for the next few hours:
- address requested changes
- implement getting two notification patterns from the env
- ensure the code passes the linting you've pushed recently

I'll write here when I'm done. You should have the edit rights, feel free to change anything for the merge. ~Please, use the new [to-be-merged-back](https://github.com/iburakov/advent-of-code-leaderboard-notifier-telegram/tree/to-be-merged-back) branch.~ Nevermind, use `main` :) Apparently I can't change the source branch of this PR, so we're stuck on main here.

iburakov committed:
Fix the fork sync conflicts

iburakov committed:
Merge remote-tracking branch 'fork-upstream/main' into main

iburakov committed:
Add autopep8 to dev requirements as it's used in `make format`

iburakov committed:
Prepare authors list comment for a merge

iburakov committed:
Rename webhook senders and related stuff to notification senders

iburakov committed:
Fix updated telegram notification sender

iburakov committed:
Implement notification pattern customization

iburakov commented:
So, after considering a few ways to enable customizing notification patterns and emojis, that's what I suggest as a flexible-yet-convenient way: 

```python
NOTIFICATION_PATTERN_EMOJIS = os.getenv("NOTIFICATION_PATTERN_EMOJIS", "🌱🎄")
NOTIFICATION_PATTERN = os.getenv(
    "NOTIFICATION_PATTERN", "Day {day}: {member} got {part_emoji} at {when}"
)
NOTIFICATION_2_PATTERN = os.getenv("NOTIFICATION_2_PATTERN", NOTIFICATION_PATTERN)
```

It seems to cover basically all discussed needs with minimal effort. Override only `NOTIFICATION_PATTERN_EMOJIS` if you need to change just the emojis, override the whole `NOTIFICATION_PATTERN` if you need to translate it, and, optionally, override `NOTIFICATION_2_PATTERN` if it must be structurally different from the first one.

iburakov committed:
Implement "after" notification pattern key

Holds "[d] hh:mm:ss" since day's puzzle publication. Arguably more
convenient than absolute :mm:ss.

iburakov commented:
A friend of mine also suggested to display the time as "[d] hh:mm:ss" since moment of corresponding puzzle publication. I tried it, it's arguably more convenient than absolute :mm:ss.

```
Day 4: Alice got 🌱 after 00:11:16
Day 4: Alice got 🎄 after 00:24:34
Day 4: Bob got 🌱 after 00:58:26
Day 4: Bob got 🎄 after 01:19:11
```

Implemented it as an `after` notification pattern key. How that sounds?  FYI, I've updated the default `NOTIFICATION_PATTERN` to this.

iburakov committed:
Try to fix the lint CI workflow so it passes

iburakov committed:
Fix silly log line

iburakov committed:
Revamp the README, describe new features, update the screenshot

iburakov commented:
Done, fixed the lint CI workflow as well. Treat all changes here as suggestions that seemed reasonable to me.

Leaving everything up to you now. You should have the edit rights, so change anything as you wish, especially README.

tofran committed:
update readme and default config values

tofran committed:
change github event to pull_request_target

tofran committed:
fix: hardcode repo owner in actions

tofran committed:
Revert "fix: hardcode repo owner in actions"

This reverts commit 70d519368e8677b165ef74c3deac8c98580f1bbd.

tofran reviewed (approved):
I think it looks great. Thank you for all the work.
Also, sorry for the broken github workflow. Should be working now (at least green on this side).

Can you spot anything or can I merge?

iburakov commented:
LGTM! Glad this worked.

tofran commented:
@iburakov right now should be available at `ghcr.io/tofran/advent-of-code-leaderboard-notifier:main`
I will do a release later.

iburakov commented:
Good, thanks. Also spotted a couple of typos at the second glance (a bit late though, sorry)

---
https://github.com/tofran/advent-of-code-leaderboard-notifier/blob/e3cf8204e2b3ac1a7c9aebaabfb9050ef9e495d1/README.md?plain=1#L86
2dn -> 2nd

---
https://github.com/tofran/advent-of-code-leaderboard-notifier/blob/e3cf8204e2b3ac1a7c9aebaabfb9050ef9e495d1/README.md?plain=1#L5
"group of friends, or co-workers" - extra comma, perhaps

tofran commented:
Thank you. Fixed those and released the new version: [v1.3.0](https://github.com/tofran/advent-of-code-leaderboard-notifier/releases/tag/v1.3.0)
